### ISSUE TITLE
Support S3 server-side-encrypted buckets as artifact store (S3 or KMS)

### ISSUE THREAD

**Issue by twiechert** [argoproj/argo#2800]
# Summary

There is currently no way to pass the required parameters to the underlying S3 client. This complicates working with SSE buckets.

# Motivation

Corporate requirements to use SSE for S3 buckets.

# Proposal

Either have properties similar to:
 
![image](https://user-images.githubusercontent.com/1926750/80116030-c3c78980-8585-11ea-9f3e-e2d8d0bfad59.png)
(took presto as an example)

or transparently pass properties to the underlying S3 client.
---

**Message from the maintainers**:

If you wish to see this enhancement implemented please add a 👍 reaction to this issue! We often sort issues this way to know what to prioritize.

**Comment by alexec** [comment]
This is implemented in [s3.go](https://github.com/argoproj/argo/blob/master/workflow/artifacts/s3/s3.go) would you like to investigate?

**Comment by twiechert** [comment]
> This is implemented in [s3.go](https://github.com/argoproj/argo/blob/master/workflow/artifacts/s3/s3.go) would you like to investigate?

The only thing I see there is a secure flag which enables encryption in transit. I was talking about encryption at rest through a customer referenced KMS key or transparently managed bv AWS. 
Minio config: https://github.com/minio/minio/tree/master/docs/security

If I am overseeing something, please become a bit more specific. Thanks.

**Comment by alexec** [comment]
@twiechert do you mean that the data must be encrypted within the workflow container?

**Comment by twiechert** [comment]
No, I'm referring to data being encrypted through S3 on server-side. 

However, it does not suffice to enable this on bucket-side but additionally, the HTTP client will need to send a control-header for this purpose. Moreover, when KMS is used for encryption, the HTTP request will also need to contain the reference to the key id. 

My request was concerned with the exposure of the respective control headers. In the case of boto s3, you will find the respective parameters prefixed with SSE:

https://boto3.amazonaws.com/v1/documentation/api/1.9.46/reference/services/s3.html

**Comment by alexec** [comment]
@twiechart ok. Would you like to submit a PR to add this feature?

**Comment by toyama0919** [comment]
sse: s3 worked.
But sse: kms doesn't work.
I checked with kubeflow-1.0.1.

**Comment by suibh** [comment]
@toyama0919, how did you get this working for s3?

**Comment by jfb74** [comment]
Any update on this? I am also working in a corporate environment that requires SSE (ServerSideEncryption AES256)

**Comment by alexec** [comment]
Would anyone like to give this a go?

**Comment by alexec** [comment]
Bump. This PR needs community contribution. The core maintainers do not operate S3 with SSE buckets, and are therefore unable to develop the feature.

**Comment by kevanslumin** [comment]
@toyama0919 How did you get SSE:S3 to work? I get an Access Denied message with archiveLogs enabled when it tries to write logs to the artifact repository.

**Comment by ksudhir007** [comment]
@alexec I would like to take this up, having worked with S3 (SSE KMS) recently, I understand why adding this support is important for adoption. I am getting up to speed with argo workflows in general.

**Comment by aaronmell** [comment]
@ksudhir007 We really need this feature as its blocking implementation of workflow in an important project for us, so we put it in our current sprint.
